### PR TITLE
refactor ipsec function

### PR DESCRIPTION
**Pull request by changluyi**
# Pull Request

- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Features
image

- Bug fixes
- Docs
- Tests

## Which issue(s) this PR fixes

Fixes #(issue-number)

### CODE REVIEW

**Comment by zhangzujian** [comment]
Could you lease add an e2e test case for this feature? I'm working on [security: run as non-root user](https://github.com/kubeovn/kube-ovn/pull/4326), which may break this function.

**Comment by changluyi** [comment]
> Could you please add an e2e test case for this feature? I'm working on [security: run as non-root user](https://github.com/kubeovn/kube-ovn/pull/4326), which may break this function.

e2e case will be added

File: Makefile [hunk 1]
```diff
@@ -915,6 +915,13 @@ kind-install-kwok:
 		kubectl apply -f kwok-node.yaml; \
 	done
 
+.PHONY: kind-install-ovn-ipsec
+kind-install-ovn-ipsec: kind-load-image
+	kubectl config use-context kind-kube-ovn
+	@$(MAKE) kind-untaint-control-plane
+	sed 's/VERSION=.*/VERSION=$(VERSION)/' dist/images/install.sh | \
+		ENABLE_OVN_IPSEC=true bash
```
Comment (zhangzujian) [inline Makefile:923]:
```suggestion
.PHONY: kind-install-ovn-ipsec
kind-install-ovn-ipsec:
	@$(MAKE) ENABLE_OVN_IPSEC=true kind-install
```

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -248,12 +249,15 @@ type Controller struct {
...
-	updateBanpQueue workqueue.RateLimitingInterface
-	deleteBanpQueue workqueue.RateLimitingInterface
-	banpKeyMutex    keymutex.KeyMutex
+	banpsLister         anplister.BaselineAdminNetworkPolicyLister
+	banpsSynced         cache.InformerSynced
+	addBanpQueue        workqueue.RateLimitingInterface
+	updateBanpQueue     workqueue.RateLimitingInterface
+	deleteBanpQueue     workqueue.RateLimitingInterface
+	banpKeyMutex        keymutex.KeyMutex
+	csrLister           certListerv1.CertificateSigningRequestLister
```
Comment (oilbeater) [inline pkg/controller/controller.go:258]: need a empty line to separate.

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/controller/controller.go (hunk 2 above)
Comment (changluyi) [reply pkg/controller/controller.go:259]: 1

File: Makefile (hunk 1 above)
Comment (changluyi) [reply Makefile]: 1

**Review by oilbeater — APPROVED** [review]

File: pkg/controller/signer.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+			klog.Errorf("Unable to approve certificate for %v and signer %v: %v", csr.Name, util.SignerName, err)
+			return err
+		}
+
+		return nil
+	}
+	// From this, point we are dealing with an approved CSR
+	// Get CA in from ovn-ipsec-ca
+	caSecret, err := c.config.KubeClient.CoreV1().Secrets("kube-system").Get(context.TODO(), util.DefaultOVNIPSecCA, metav1.GetOptions{})
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:137]: need log err

File: pkg/controller/signer.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+		c.signerFailure(csr, "CAFailure",
+			fmt.Sprintf("Could not get CA certificate and key: %v", err))
+		return err
+	}
+
+	// Decode the certificate request from PEM format.
+	certReq, err := decodeCertificateRequest(csr.Spec.Request)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:145]: need log err

File: pkg/controller/signer.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+		// malformed CSR rather than an issue with the controller.
+		if err := c.updateCSRStatusConditions(csr, "CSRDecodeFailure", fmt.Sprintf("Could not decode Certificate Request: %v", err)); err != nil {
+			klog.Error(err)
+		}
+		return nil
+	}
+
+	// Decode the CA certificate from PEM format.
+	caCert, err := decodeCertificate(caSecret.Data["cacert"])
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:156]: need log err

File: pkg/controller/signer.go [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+		c.signerFailure(csr, "CorruptCACert",
+			fmt.Sprintf("Unable to decode CA certificate for %v: %v", util.SignerName, err))
+		return nil
+	}
+
+	caKey, err := decodePrivateKey(caSecret.Data["cakey"])
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:163]: need log err

File: pkg/controller/signer.go [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+		c.signerFailure(csr, "CorruptCAKey",
+			fmt.Sprintf("Unable to decode CA private key for %v: %v", util.SignerName, err))
+		return nil
+	}
+
+	// Create a new certificate using the certificate template and certificate.
+	// We can then sign this using the CA.
+	signedCert, err := signCSR(newCertificateTemplate(certReq), certReq.PublicKey, caCert, caKey)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:172]: need log err

File: pkg/controller/signer.go [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+		c.signerFailure(csr, "SigningFailure",
+			fmt.Sprintf("Unable to sign certificate for %v and signer %v: %v", csr.Name, util.SignerName, err))
+		return nil
+	}
+
+	// Encode the certificate into PEM format and add to the status of the CSR
+	csr.Status.Certificate, err = encodeCertificates(signedCert)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:180]: need log err

File: pkg/controller/signer.go [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+		c.signerFailure(csr, "EncodeFailure",
+			fmt.Sprintf("Could not encode certificate: %v", err))
+		return nil
+	}
+
+	if err := c.updateCsrStatus(csr); err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:186]: need log err

File: pkg/controller/signer.go [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+// Update the status conditions on the CSR object
+func (c *Controller) updateCSRStatusConditions(csr *csrv1.CertificateSigningRequest, reason, message string) error {
+	csr.Status.Conditions = append(csr.Status.Conditions, csrv1.CertificateSigningRequestCondition{
+		Type:    csrv1.CertificateFailed,
+		Status:  "True",
+		Reason:  reason,
+		Message: message,
+	})
+
+	if err := c.updateCsrStatus(csr); err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:212]: need log err

File: pkg/controller/signer.go [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+		if c.Type == csrv1.CertificateDenied {
+			denied = true
+		}
+	}
+	return
+}
+
+func newCertificateTemplate(certReq *x509.CertificateRequest) *x509.Certificate {
+	serialNumber, err := rand.Int(rand.Reader, big.NewInt(1<<62))
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:249]: need log err

File: pkg/controller/signer.go [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+		DNSNames:              certReq.DNSNames,
+		BasicConstraintsValid: true,
+	}
+
+	return template
+}
+
+func signCSR(template *x509.Certificate, requestKey c.PublicKey, issuer *x509.Certificate, issuerKey c.PrivateKey) (*x509.Certificate, error) {
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, issuer, requestKey, issuerKey)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:271]: need log err

File: pkg/controller/signer.go [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+		return nil, err
+	}
+	certs, err := x509.ParseCertificates(derBytes)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:275]: need log err

File: pkg/controller/signer.go [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+	}
+	if len(certs) != 1 {
+		return nil, errors.New("Expected a single certificate")
+	}
+	return certs[0], nil
+}
+
+func decodeCertificateRequest(pemBytes []byte) (*x509.CertificateRequest, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
```
Comment (bobz965) [inline pkg/controller/signer.go:286]: need log err

File: pkg/controller/signer.go [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+		return nil, err
+	}
+
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+func decodeCertificate(pemBytes []byte) (*x509.Certificate, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != "CERTIFICATE" {
+		err := errors.New("PEM block type must be CERTIFICATE")
```
Comment (bobz965) [inline pkg/controller/signer.go:297]: need log err

File: pkg/controller/signer.go [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+		return nil, err
+	}
+
+	return x509.ParseCertificate(block.Bytes)
+}
+
+func decodePrivateKey(pemBytes []byte) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil || block.Type != "PRIVATE KEY" {
+		fmt.Println(block.Type)
```
Comment (bobz965) [inline pkg/controller/signer.go:307]: need log err

File: pkg/controller/signer.go [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+		err := errors.New("PEM block type must be PRIVATE KEY")
+		return nil, err
+	}
+
+	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:313]: need log err

File: pkg/controller/signer.go [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+		return nil, err
+	}
+
+	return rsaKey, nil
+}
+
+func encodeCertificates(certs ...*x509.Certificate) ([]byte, error) {
+	b := bytes.Buffer{}
+	for _, cert := range certs {
+		if err := pem.Encode(&b, &pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}); err != nil {
```
Comment (bobz965) [inline pkg/controller/signer.go:329]: need log err

File: pkg/daemon/ipsec.go [hunk 19]
```diff
@@ -0,0 +1,384 @@
...
+	ipsecCACertPath  = ipsecKeyDir + "ipsec-cacert.pem"
+	ipsecCertPath    = ipsecKeyDir + "ipsec-cert.pem"
+
+	expireTime = 365 * 24 * time.Hour
+)
+
+func getOVSSystemID() (string, error) {
+	cmd := exec.Command("ovs-vsctl", "--retry", "-t", "60", "get", "Open_vSwitch", ".", "external-ids:system-id")
+	output, err := cmd.Output()
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:35]: need log err

File: pkg/daemon/ipsec.go [hunk 20]
```diff
@@ -0,0 +1,384 @@
...
+	if systemID == "" {
+		return "", errors.New("empty system-id")
+	}
+
+	return systemID, nil
+}
+
+func checkCertExpired() (bool, error) {
+	certBytes, err := os.ReadFile(ipsecCertPath)
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:50]: need log err

File: pkg/daemon/ipsec.go [hunk 21]
```diff
@@ -0,0 +1,384 @@
...
+		return false, fmt.Errorf("failed to read certificate: %w", err)
+	}
+
+	block, _ := pem.Decode(certBytes)
+	if block == nil {
+		return false, errors.New("failed to decode PEM block containing certificate")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:60]: need log err

File: pkg/daemon/ipsec.go [hunk 22]
```diff
@@ -0,0 +1,384 @@
...
+	if time.Since(cert.NotBefore) > expireTime {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func generateCSRCode() ([]byte, error) {
+	cn, err := getOVSSystemID()
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:73]: need log err

File: pkg/daemon/ipsec.go [hunk 23]
```diff
@@ -0,0 +1,384 @@
...
+		return nil, err
+	}
+
+	klog.Infof("ovs system id: %s", cn)
+	cmd := exec.Command("openssl", "genrsa", "-out", ipsecPrivKeyPath, "2048")
+	err = cmd.Run()
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:80]: need log err

File: pkg/daemon/ipsec.go [hunk 24]
```diff
@@ -0,0 +1,384 @@
...
+		return nil, err
+	}
+
+	_, err = os.Stat(ipsecPrivKeyPath)
+	if err != nil {
+		if os.IsNotExist(err) {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:86]: need log err

File: pkg/daemon/ipsec.go [hunk 25]
```diff
@@ -0,0 +1,384 @@
...
+	}
+
+	cmd = exec.Command("openssl", "req", "-new", "-text",
+		"-extensions", "v3_req",
+		"-addext", "subjectAltName = DNS:"+cn,
+		"-subj", fmt.Sprintf("/C=CN/O=kubeovn/OU=kind/CN=%s", cn),
+		"-key", ipsecPrivKeyPath,
+		"-out", ipsecReqPath) // #nosec
+	err = cmd.Run()
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:99]: need log err

File: pkg/daemon/ipsec.go [hunk 26]
```diff
@@ -0,0 +1,384 @@
...
+		return nil, err
+	}
+
+	csrBytes, err := os.ReadFile(ipsecReqPath)
+	if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:104]: need log err

File: pkg/daemon/ipsec.go [hunk 27]
```diff
@@ -0,0 +1,384 @@
...
+		Spec: v1.CertificateSigningRequestSpec{
+			Request:    csrBytes,
+			SignerName: util.SignerName,
+			Usages: []v1.KeyUsage{
+				v1.UsageIPsecTunnel,
+			},
+		},
+	}
+
+	if _, err := c.config.KubeClient.CertificatesV1().CertificateSigningRequests().Create(context.Background(), csr, metav1.CreateOptions{}); err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:129]: need log err

File: pkg/daemon/ipsec.go [hunk 28]
```diff
@@ -0,0 +1,384 @@
...
+		return err
+	}
+
+	// Wait until the certificate signing request has been signed.
+	var certificateStr string
+	counter := 0
+	for {
+		csr, err := c.config.KubeClient.CertificatesV1().CertificateSigningRequests().Get(context.Background(), csr.Name, metav1.GetOptions{})
+		if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:138]: need log err

File: pkg/daemon/ipsec.go [hunk 29]
```diff
@@ -0,0 +1,384 @@
...
+	// clear /etc/openvswitch/keys/ipsec-cacert.pem
+	cmd := exec.Command("rm", "-f", "/etc/openvswitch/keys/ipsec-cacert.pem")
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func initIPSecKeysDir() error {
+	if err := os.MkdirAll(ipsecKeyDir, 0o755); err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:229]: need log err

File: pkg/daemon/ipsec.go [hunk 30]
```diff
@@ -0,0 +1,384 @@
...
+func (c *Controller) ManageIPSecKeys() error {
+	_, err := os.Stat(ipsecCertPath)
+	if os.IsNotExist(err) {
+		if err := c.CreateIPSecKeys(); err != nil {
+			klog.Errorf("create ipsec keys error: %v", err)
+			return err
+		}
+	} else {
+		checkCertExpired, err := checkCertExpired()
+		if err != nil {
```
Comment (bobz965) [inline pkg/daemon/ipsec.go:260]: need log err

**Comment by bobz965** [comment]
@changluyi 大佬，我感觉有些err log 还是就近打印更好些，后面有空麻烦补充下

**Pull request merged**